src/esp/gfx_batch/Hbao.cpp

```diff
@@ -0,0 +1,1069 @@
+    } else if (configuration.flags() & HbaoFlag::LayeredGeometryShader) {
+      layered = HbaoCalcShader::Layered::GeometryShader;
+      textureArrayLayer = false;
+    }
+    state_->hbao2CalcShader = HbaoCalcShader{
+        /*deinterleaved*/ true, /*blur*/ false, layered, textureArrayLayer};
+    state_->hbao2CalcBlurShader = HbaoCalcShader{
+        /*deinterleaved*/ true, /*blur*/ true, layered, textureArrayLayer};
+  }
+
+  state_->triangle.setCount(3);
+  state_->triangleLayered.setCount(3 * HbaoRandomNumElements);
+  state_->configuration = configuration;
+}
+
+Magnum::Vector2i Hbao::getFrameBufferSize() const {
+  return state_->configuration.size();
+}
+
+namespace {
+
+/**
+ * Populate uniform data with appropriate values from configuration. Should only
+ * be performed when configuration changes.
+ */
+void prepareHbaoData(
+    const HbaoConfiguration& configuration,
+    const Mn::Matrix4& projection,
+    HbaoUniformData& uniformData,
+    Mn::GL::Buffer& uniform,
+    Cr::Containers::StaticArrayView<HbaoRandomNumElements * MaxSamples,
+                                    const Mn::Vector4> random) {
+  /* Radius */
+  const Mn::Float meters2viewspace = 1.0f;
+  const Mn::Float r = configuration.radius() * meters2viewspace;
+  uniformData.r2 = r * r;
+  uniformData.negInvR2 = -1.0f / uniformData.r2;
+  const Mn::Float projectionScale =
+      (uniformData.projOrtho != 0
+           ? configuration.size().y() / uniformData.projInfo[1]  // ortho
+           // For perspective, projection[0][0] is 1/tan(fov/2)
+           : 0.5f * configuration.size().y() * projection[0][0]);  // persp
+  uniformData.radiusToScreen = r * 0.5f * projectionScale;
+
+  /* AO */
+  uniformData.powExponent = Mn::Math::max(configuration.intensity(), 0.0f);
+  uniformData.nDotVBias = Mn::Math::clamp(configuration.bias(), 0.0f, 1.0f);
+  uniformData.aoMultiplier = 1.0f / (1.0f - uniformData.nDotVBias);
+
+  /* Resolution */
+  const Mn::Vector2i quarterSize = (configuration.size() + Mn::Vector2i{3}) / 4;
+  uniformData.invQuarterResolution =
+      Mn::Vector2{1.0f} / Mn::Vector2{quarterSize};
+  uniformData.invFullResolution =
+      Mn::Vector2{1.0f} / Mn::Vector2{configuration.size()};
+
+  if (configuration.flags() &
+      (HbaoFlag::LayeredGeometryShader | HbaoFlag::LayeredImageLoadStore))
+    for (Mn::Int i = 0; i != HbaoRandomNumElements; ++i) {
+      uniformData.float2Offsets[i] = {
+          Mn::Float(i % 4) + 0.5f,
+          // NOLINTNEXTLINE(bugprone-integer-division). This is deliberate
+          Mn::Float(i / 4) + 0.5f,
+          0.0f,
+          0.0f,
+      };
+      uniformData.jitters[i] = random[i];
+    }
+
+  uniform.setSubData(0, {&uniformData, 1});
+}
+
+Mn::Vector4 buildClipInfo(const Mn::Matrix4& projectionMatrix,
+                          bool orthographic) {
+  // TODO this still looks like there's some better way this should be
+  // extracted.  Definitely change to use near and far calcs upstream when
+  // available!!
```

Comment:
👍

Reply:
Might want to reference this commit in the TODO: https://github.com/mosra/magnum/commit/0d31f7461b31698ea5bf92ec66ff5056a6ad7360